libraries/botbuilder-core/src/activityHandler.ts

```diff
@@ -231,50 +231,191 @@ export class ActivityHandler {
-                        if (context.activity.name === 'tokens/response') {
-                            await this.handle(context, 'TokenResponseEvent', runDialogs);
-                        } else {
-                            await runDialogs();
-                        }
-                    });
-                    break;
-                default:
-                // handler for unknown or unhandled types
-                    await this.handle(context, 'UnrecognizedActivityType', runDialogs);
-                    break;
-            }
+            await this.dispatchActivity(context, next);
+        });
+    }
+
+    /**
+     * Overwrite this method to use different dispatching logic after completing all `onTurn`-registered handlers.
+     * @remarks
+     * The default logic is below:
+     * ```ts
+     *      switch (context.activity.type) {
+     *          case ActivityTypes.Message:
+     *              await this.handleMessageActivity(context, next);
+     *              break;
+     *          case ActivityTypes.ConversationUpdate:
+     *              await this.handleConversationUpdateActivity(context, next);
+     *              break;
+     *          case ActivityTypes.MessageReaction:
+     *              await this.handleMessageReactionActivity(context, next);
+     *              break;
+     *          case ActivityTypes.Event:
+     *              await this.handleEventActivity(context, next);
+     *              break;
+     *          default:
+     *              // handler for unknown or unhandled types
+     *              await this.handleUnrecognizedActivity(context, next);
+     *              break;
+     *      }
+     * ```
+     * @param context TurnContext A TurnContext representing an incoming Activity from an Adapter
+     * @param next () => Promise<void>
+     */
+    async dispatchActivity(context: TurnContext, next: () => Promise<void>): Promise<void> {
+        switch (context.activity.type) {
+            case ActivityTypes.Message:
+                await this.handleMessageActivity(context, next);
+                break;
+            case ActivityTypes.ConversationUpdate:
+                await this.handleConversationUpdateActivity(context, next);
+                break;
+            case ActivityTypes.MessageReaction:
+                await this.handleMessageReactionActivity(context, next);
+                break;
+            case ActivityTypes.Event:
+                await this.handleEventActivity(context, next);
+                break;
+            default:
+            // handler for unknown or unhandled types
+                await this.handleUnrecognizedActivity(context, next);
+                break;
+        }
+    }
+
+    /**
+     * 
+     * @param context TurnContext A TurnContext representing an incoming Activity from an Adapter
+     * @param next () => Promise<void>
+     */
+    protected async handleMessageActivity(context: TurnContext, next: () => Promise<void>): Promise<void> {
```

Comment:
this is now _very_ close to the C# onMessageActivity ... perhaps in the next iteration we could unify further
(FWIW the introduction of functions here is massively better factoring - much clear to read)